UI: Breaking `ControlledTree` changes (#2186)
imodeljs/imodeljs: docs/changehistory/NextVersion.md

#### `ControlledTree` API Changes

`ControlledTree` component has received the following breaking changes:

* The component now takes `TreeModel` rather than `VisibleTreeNodes` as a prop to avoid requiring consumers to manage `VisibleTreeNodes` object. As a result, the `useVisibleTreeNodes` hook was replaced with `useTreeModel` hook. Typical migration:

  **Before:**

  ```tsx
  const visibleNodes = useVisibleTreeNodes(modelSource);
  return <ControlledTree visibleNodes={visibleNodes} {...otherProps} />;
  ```

  **After:**

  ```tsx
  const treeModel = useTreeModel(modelSource);
  return <ControlledTree model={treeModel} {...otherProps} />;
  ```

* Name of the `treeEvents` prop was changed to `eventsHandler` to make it clearer. Typical migration:

  **Before:**

  ```tsx
  return <ControlledTree treeEvents={eventsHandler} {...otherProps} />;
  ```

  **After:**

  ```tsx
  return <ControlledTree eventsHandler={eventsHandler} {...otherProps} />;
  ```

> * Made the props `width` and `height` required. Previously they were optional and forced us to use non-optimal approach when not provided. Now it's up to the consumer to tell the size of the component. Typical migration:

Reviewer: We should probably stick to present tenses for consistency.

```suggestion
* `width` and `height` properties are now required. Previously they were optional and forced us to use non-optimal approach when not provided. Now it's up to the consumer to tell the size of the component. Typical migration:
```

The method `getFloatingWidgetContainerIds()` has been added to FrontstageDef to retrieve the Ids for all floating widget containers for the active frontstage as specified by the `frontstageDef`. These ids can be used to query the size of the floating container via `frontstageDef.getFloatingWidgetContainerBounds`. The method `frontstageDef.setFloatingWidgetContainerBounds` can then be used to set the size and position of a floating widget container.

#### `ControlledTree` API Changes

> There were some breaking API changes related to `ControlledTree` component:

Reviewer: ```suggestion
`ControlledTree` component has received the following breaking changes:
```

> * The component now takes `TreeModel` rather than `VisibleTreeNodes` as a prop to avoid requiring consumers to manage `VisibleTreeNodes` object. As a result, the `useVisibleTreeNodes` hook was replaced with `useTreeModel` hook. Typical mitigation:

Reviewer: Instead of `Typical mitigation` let's try `Typical fix` or `Typical migration`

Author: Changed to "migration"

  **Before:**

  ```tsx
  const visibleNodes = useVisibleTreeNodes(modelSource);
  return (
    <ControlledTree {...otherProps} visibleNodes={visibleNodes} />
  );

>   ```

Reviewer: The code fragment could be shorter:
```tsx
const visibleNodes = useVisibleTreeNodes(modelSource);
return <ControlledTree visibleNodes={visibleNodes} {...otherProps} />
```

Also, maybe put the relevant prop before the rest.

Author: Done, thanks

  **After:**

  ```tsx
  const treeModel = useTreeModel(modelSource);
  return (
    <ControlledTree {...otherProps} model={treeModel} />
  );
  ```

* Name of the `treeEvents` prop was changed to `eventsHandler` to make it clearer. Typical mitigation:

  **Before:**

  ```tsx
  return (
    <ControlledTree {...otherProps} treeEvents={eventsHandler} />
  );
  ```

  **After:**

  ```tsx
  return (
    <ControlledTree {...otherProps} eventsHandler={eventsHandler} />
  );
  ```

> * Made the props `width` and `height` required. Previously they were optional and forced us to use non-optimal approach when not provided. Now it's up to the consumer to tell the size of the component. Typical mitigation:

Reviewer: Developers almost certainly aren't aware that `width` and `height` properties even existed as they only have been added recently. Also, we probably shouldn't try to sell people on removal of a feature when it was flawed from the start.

```suggestion
* `width` and `height` properties are now required. Typical fix:
```

Author: I don't see a problem with us explaining why we removed a feature (flawed).

  **Before:**

  ```tsx
  return (
    <ControlledTree {...props} />
  );
  ```

  **After:**

  ```tsx
  const width = 100;
  const height = 100;
  return (
    <ControlledTree {...props} width={width} height={height} />
  );
  ```

>   Or, the size may be determined dynamically. The below example uses an `AutoSizer` component from `react-virtualized-auto-sizer` package for that:

Reviewer: We should direct people to use [`ResizeObserver`](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) instead.

Author: I don't mind us referencing `ResizeObserver` as a possible solution, but I don't want to give an example with it, because it would be much larger. What do you think about rephrasing the comment like:
```markdown
Or, the size may be determined dynamically using [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) or some third-party library. The below example uses an `AutoSizer` component from `react-virtualized-auto-sizer` package for that:
```
?

Author: Changed to
```markdown
`width` and `height` props may be calculated dynamically using [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) API.
```
and removed the example - let's leave it up to consumers to decide how to come up with width and height as it's fairly straightforward to find and anything we suggest may become out of date quickly.

Reviewer: Agree, thanks.
